crates/types/src/state_mut.rs

```diff
@@ -21,3 +23,89 @@ pub struct ExternalStateMutation {
     pub version: Option<String>,
     pub state: HashMap<Bytes, Bytes>,
 }
+
+/// # StateMutationVersion
+///
+/// This type represents a user state version. This implementation hashes canonically the raw key-value
+/// and hands out an opaque string representation of that version, to be used for exact comparisons.
+///
+#[derive(Eq, PartialEq, Debug, Ord, PartialOrd)]
+pub struct StateMutationVersion(String);
+
+impl StateMutationVersion {
+    pub fn from_raw<S: Into<String>>(raw: S) -> StateMutationVersion {
+        StateMutationVersion(raw.into())
+    }
+
+    pub fn from_user_state(state: &[(Bytes, Bytes)]) -> StateMutationVersion {
+        let mut kvs: Vec<_> = state.iter().collect();
+        kvs.sort_by_key(|(k, _)| k);
+
+        let mut hasher = Sha256::new();
+        for (i, (k, v)) in kvs.iter().enumerate() {
+            hasher.update(i.to_be_bytes());
+            hasher.update([0x1u8]);
```

Comment:
Curious to learn why these constants are important for the hash creation. Would otherwise be a higher risk of collisions?